### ISSUE TITLE
A11y audits: Wall of text / false positives for contrast ratio violations

### ISSUE THREAD

**Issue by pavelfeldman** [GoogleChrome/lighthouse#1615]
Consider delivering a screenshot with questionable areas outlined. If that is not feasible due to the false positives, consider dropping the audit. I am not aware of a good CSS-based implementation of this due to bg images. A more reliable implementation would render a screenshot with and without the text and then would compare those.

![contrast_ratio](https://cloud.githubusercontent.com/assets/883973/22571660/6bedc49c-e955-11e6-908c-e1e1ee90fc12.png)

**Comment by rossPatton** [comment]
i'm also having issues with this. this check does not appear reliable or consistent

1 ) my current site passes WAVE and google accessibility audit tools here but not when using lighthouse and

2 ) elements are passing with less contrast than some elements that are failing with higher contrast. see attached

it basically feels like it's impossible to pass this test without having a black and white page. it should be removed if it's broken. or preferably fixed to be consistent with other tools

bigger text with darker background color, fails test:
![screen shot 2017-03-08 at 6 36 09 pm](https://cloud.githubusercontent.com/assets/2379901/23729104/1ef00378-042e-11e7-9301-2b70f95be6da.png)

smaller text, with lighter background color, passes test:
![screen shot 2017-03-08 at 6 36 18 pm](https://cloud.githubusercontent.com/assets/2379901/23729106/23a1a520-042e-11e7-9f1e-18e0129831c1.png)

and for comparisons sake, what the above look like when running chrome in high contrast mode

again, this passes
![screen shot 2017-03-08 at 6 39 28 pm](https://cloud.githubusercontent.com/assets/2379901/23729223/9fd7e06e-042e-11e7-9597-794b5af639a9.png)

but this doesn't
![screen shot 2017-03-08 at 6 39 34 pm](https://cloud.githubusercontent.com/assets/2379901/23729227/a913d1f6-042e-11e7-9f2a-02dd8635c2a6.png)

**Comment by ebidel** [comment]
cc @robdodson

**Comment by robdodson** [comment]
I'm happy to look into this but it may take a bit cuz I'm traveling to Sydney later today.

@rossPatton are you able to share a URL that I can test against?

**Comment by alice** [comment]
There are two issues here:

1. axe-core returns a query selector for each node violating each rule. I have work in progress to allow it to return a live element for the top frame, which should help with this in general.
2. Potential false positives for the color contrast rule. As @robdodson said, live examples would help us and the axe-core developers dig into this issue.

**Comment by robdodson** [comment]
@paulirish you mentioned on Hangouts you might not need the live element references anymore, can you fill us in on that?

**Comment by paulirish** [comment]
@robdodson yeahhhhh.. here's the deal. cc @alice 

so dgozman started working on element revealing last week. He got pretty close to done and then noticed that we were using nodeId from a DOM that has since been refreshed. So those nodes are long gone. :)

We are now thinking of just using the node path to resolve these nodes, rather than figure out how to always avoid reloads.

Turns out DevTools already has a trick like this, when it reselects the node you had inspected right before the reload. Here's what happens

1. Reload is initiated
1. `Elements.ElementsPanel_documentUpdated` runs and grabs `_selectedNodeOnReset`
1. It gets `node.path()` which is [implemented here](https://github.com/ChromeDevTools/devtools-frontend/blob/0cc682f1be6c98eb01f47276ad2954fd9ace5653/front_end/sdk/DOMModel.js#L522-L544) and returns a string like `"1,HTML,1,BODY,0,DIV,0,H1"`
1. In the new DOM, that path is thrown to [`DOM.pushNodeByPathToFrontend`](https://chromedevtools.github.io/debugger-protocol-viewer/tot/DOM/#method-pushNodeByPathToFrontend) and a nodeId is returned
1. It gets handled in `onNodeRestored` and is re-inspected there.

---

So, if this makes sense to ya'll, it probably makes sense to get same path format for these axe nodes. 

Not sure where it's best to do that, but assuming that Axe isn't interested in this proprietary node path format (heh), I guess returning real live elements actually _does_ makes sense. We can then generate the node path afterwards.

hows that sound?

**Comment by rossPatton** [comment]
@robdodson I can provide one next week once we go live, thanks for offering to look into it

**Comment by alice** [comment]
@rossPatton Friendly bump for URL demonstrating problem :smile:

**Comment by rossPatton** [comment]
@alice just went live: http://www.iodine.com/drug/zoloft

**Comment by alice** [comment]
As above; found the issue in axe-core which is causing the false positives on color contrast here.

**Comment by alice** [comment]
The false positives have now been fixed in axe-core: https://github.com/dequelabs/axe-core/pull/272

**Comment by ojanvafai** [comment]
Don't know if this is the same problem as the one that was just fixed, but here's a reduced case:

```
<div style="position: absolute; bottom: 0; background-color: black; color: white;">asdf</div>
```

It's not the position absolute. If I use flexbox to put it at the bottom instead, I also get the false positive. If I remove the "bottom: 0", then it passes.

**Comment by dandv** [comment]
Lyft has also had this problem, CC @mohsen1

**Comment by robdodson** [comment]
@dandv looking at that report I can't seem to find the elements on the page, do I need to be signed in? Can you share a screenshot of them?

I don't think the change to axe-core has been released yet, cc @marcysutton

**Comment by mohsen1** [comment]
The pickup button is marked as low contrast 

![img_1453](https://cloud.githubusercontent.com/assets/543633/24889460/dd6b6e5a-1e1e-11e7-9efd-e8eb86037047.PNG)

**Comment by alice** [comment]
@moshen1 That definitely looks like the same bug.

**Comment by alice** [comment]
New axe-core is out now. This also has the option to return real element references via the `elementRef` option.

**Comment by ebidel** [comment]
Woot! @alice or @robdodson is updating us to the new version and utilizing the element references something you'd like to help with? I don't think there's a huge rush to modernize if you're busy.

**Comment by alice** [comment]
I think @paulirish is on it already, but I'm around if there are any questions :grin:

**Comment by rossPatton** [comment]
has this fix been merged into lighthouse?

**Comment by robdodson** [comment]
Yep https://github.com/GoogleChrome/lighthouse/pull/2090 contained the fix.

@ebidel should we close this?

**Comment by ebidel** [comment]
Done!

**Comment by patrickhulce** [comment]
@rossPatton yes version 2.0.0 of Lighthouse has axe 2.2.0 and re-running on that URL you provided seems to have just 1 valid failure now for an [orange background](http://leaverou.github.io/contrast-ratio/#%23ff5722-on-white).

**Comment by rossPatton** [comment]
lighthouse 2.0 is very 👍
